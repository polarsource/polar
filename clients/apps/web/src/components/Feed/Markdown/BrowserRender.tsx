import Markdown from 'markdown-to-jsx'
import dynamic from 'next/dynamic'
import { Skeleton } from 'polarkit/components/ui/skeleton'
import { createContext, useContext } from 'react'
import BrowserAd from './Ad/BrowserAd'
import Embed from './Embed/BrowserEmbed'
import Iframe from './Iframe/BrowserIframe'
import { ImageOverlay } from './Img/ImageOverlay'
import Paywall from './Paywall/Paywall'
import BrowserPoll from './Poll/BrowserPoll'
import Poll from './Poll/Poll'
import SubscribeNow from './SubscribeNow/SubscribeNow'
import {
  RenderArticle,
  markdownOpts,
  wrapStrictCreateElement,
} from './markdown'

const BrowserMermaid = dynamic(() => import('./Mermaid/BrowserMermaid'), {
  ssr: false,
  loading: () => (
    <Skeleton className="w-full p-8 font-mono text-sm">
      Initializing Mermaid Renderer...
    </Skeleton>
  ),
})

// Dynamically load the SyntaxHighlighter (heavily reduces bundle sizes)
//
// While loading (and SSR), render a placeholder without syntax highlighting
const SyntaxHighlighterContext = createContext('')
const BrowserSyntaxHighlighter = dynamic(
  () => import('./SyntaxHighlighter/BrowserSyntaxHighlighter'),
  {
    ssr: false,
    loading: () => <BrowserSyntaxHighlighterLoading />,
  },
)

const BrowserSyntaxHighlighterLoading = () => {
  const value = useContext(SyntaxHighlighterContext)
  return (
    <pre className="w-full">
      <code className="text-black dark:text-white">{value}</code>
    </pre>
  )
}

export const opts = {
  ...markdownOpts,

  overrides: {
    ...markdownOpts.overrides,

    // browser overrides
    poll: (args: any) => <Poll {...args} renderer={BrowserPoll} />,
    Paywall: (args: any) => <Paywall {...args} />,
    SubscribeNow: (args: any) => <SubscribeNow {...args} />,
    embed: (args: any) => <Embed {...args} />,
    iframe: (args: any) => <Iframe {...args} />,
    img: (args: any) => <ImageOverlay {...args} />,
    pre: (args: any) => {
      if (args.children.type === 'code') {
        const language = args.children.props.className?.replace('lang-', '')
        if (language === 'mermaid') {
          return (
            <BrowserMermaid graphDefinition={args.children.props.children} />
          )
        }
        return (
          <SyntaxHighlighterContext.Provider
            value={args.children.props.children}
          >
            <BrowserSyntaxHighlighter
              language={language}
              {...args.children.props}
            />
          </SyntaxHighlighterContext.Provider>
        )
      }
      return <></>
    },
    Ad: (args: any) => <BrowserAd {...args} />,
  },
} as const

export const previewOpts = {
  ...markdownOpts,
  overrides: {
    ...markdownOpts.overrides,

    poll: () => <></>,
    Paywall: () => <></>,
    SubscribeNow: () => <></>,
    embed: () => <></>,
    iframe: () => <></>,
    pre: () => <></>,
  },
} as const

const BrowserRender = ({
  article,
  showPaywalledContent,
  isSubscriber,
  paidArticlesBenefitName,
}: {
  article: RenderArticle
  showPaywalledContent: boolean
  isSubscriber: boolean
  paidArticlesBenefitName?: string
}) => {
  return (
    <Markdown
      options={{
        ...opts,
        createElement: wrapStrictCreateElement({
          article,
          showPaywalledContent,
          isSubscriber,
          paidArticlesBenefitName,
          extraAllowedCustomComponents: Object.keys(opts.overrides),
        }),
      }}
    >
      {article.body}
    </Markdown>
  )
}

export const AbbreviatedBrowserRender = ({
  article,
  showPaywalledContent,
}: {
  article: RenderArticle
  showPaywalledContent?: boolean
}) => {
  return (
    <Markdown
      options={{
        ...previewOpts,
        createElement: wrapStrictCreateElement({
          article,
          showPaywalledContent,
          isSubscriber: true, // Do not show <SubscribeNow /> in abbreviations
        }),
      }}
    >
      {article.body.substring(0, 500).split('\n').slice(0, 4).join('\n')}
    </Markdown>
  )
}

export default BrowserRender
