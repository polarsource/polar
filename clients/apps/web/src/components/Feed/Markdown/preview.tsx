import { Article } from '@polar-sh/sdk'
import { decode } from 'html-entities'
import Markdown from 'markdown-to-jsx'
import { markdownOpts, wrapStrictCreateElement } from './markdown'

export const previewOpts = {
  ...markdownOpts,
  overrides: {
    ...markdownOpts.overrides,
    p: (args: any) => <>{args.children} </>, // Note the space
    div: (args: any) => <>{args.children} </>, // Note the space
    img: (args: any) => <></>,
    a: (args: any) => <>{args.children}</>,
    strong: (args: any) => <>{args.children}</>,
    em: (args: any) => <>{args.children}</>,
  },
} as const

// Text only renderer. Used to generate contents for og:description and similar
export default function PreviewText(props: { article: Article }) {
  return (
    <Markdown
      options={{
        ...previewOpts,
        createElement: wrapStrictCreateElement({
          article: props.article,
          showPaywalledContent: false,
          extraAllowedCustomComponents: [
            'p',
            'div',
            'img',
            'a',
            'strong',
            'em',
          ],

          // Default override that removes all elements
          defaultOverride: (args: any) => <>{args.children}</>,
        }),
        wrapper: (args: any) => <>{args.children}</>,
      }}
    >
      {props.article.body.substring(0, 500)}
    </Markdown>
  )
}

export function UnescapeText(str: string): string {
  return decode(str, { scope: 'strict' }).trim()
}
