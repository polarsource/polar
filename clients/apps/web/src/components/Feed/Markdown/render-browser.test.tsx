import { article } from 'polarkit/testdata'

import { PolarQueryClientProvider } from '@/app/providers'
import '@testing-library/jest-dom'
import { act, render } from '@testing-library/react'
import Markdown from '@zegl/markdown-to-jsx'
import { opts } from './BrowserRender'
import { RenderArticle, wrapStrictCreateElement } from './markdown'
import { polarPostUpsellAccess } from './testdata/polarPostUpsellAccess'

const TestRenderer = (props: { article: RenderArticle }) => {
  return (
    <PolarQueryClientProvider>
      <Markdown
        options={{
          ...opts,
          createElement: wrapStrictCreateElement({
            article: props.article,
            extraAllowedCustomComponents: Object.keys(opts.overrides),
          }),
        }}
      >
        {props.article.body}
      </Markdown>
    </PolarQueryClientProvider>
  )
}

test('code', async () => {
  let asFragment

  await act(() => {
    const component = render(
      <TestRenderer
        article={{
          ...article,
          body: `
 
code:::

\`\`\`go
func main() {
}
\`\`\`

`,
        }}
      />,
    )
    asFragment = component.asFragment
  })

  // @ts-ignore
  expect(asFragment()).toMatchSnapshot()
})

test('XSS', async () => {
  let asFragment

  await act(() => {
    const component = render(
      <TestRenderer
        article={{
          ...article,
          body: `
 
<p>
hello
</p>

<p onload="alert(1)">
what
</p>
    
<table>
    <thead>
        <tr><th>TH</th></tr>
    </thead>
    <tbody onload="alert(2)">
        <tr onload="alert(3)"><td onload="alert(4)">TD</td></tr>
        <tr><td><p onload="alert(1)"><img src="not-sanitized" onclick="xxx" /> <a href="javascript:alert(1)">clickme</a> <span onclick="xxx">clickmep</span></p></td><td onclick="x">TD</td></tr>
    </tbody>
</table>

`,
        }}
      />,
    )
    asFragment = component.asFragment
  })

  // @ts-ignore
  expect(asFragment()).toMatchSnapshot()
})

test('basic', async () => {
  let asFragment

  await act(() => {
    const component = render(
      <TestRenderer
        article={{
          ...article,
          body: `
# h1

## h2

### h3

Hello **world**!

[Polar](https://polar.sh/)

<Paywall></Paywall>

This is a normal **block** of text [Polar](https://polar.sh/) with _various_ formatting.
And here it continues in the same block.

This is a different block.  
With a linebreak! (double whitespace)

> This is a quoute!

`,
        }}
      />,
    )
    asFragment = component.asFragment
  })

  // @ts-ignore
  expect(asFragment()).toMatchSnapshot()
})

test('posts', async () => {
  let asFragment

  await act(() => {
    const component = render(
      <TestRenderer
        article={{
          ...article,
          body: polarPostUpsellAccess,
        }}
      />,
    )
    asFragment = component.asFragment
  })

  // @ts-ignore
  expect(asFragment()).toMatchSnapshot()
})

test('mermaid', async () => {
  let asFragment

  await act(() => {
    const component = render(
      <TestRenderer
        article={{
          ...article,
          body: `

we love charts

\`\`\`mermaid
pie title Pets adopted by volunteers
    "Dogs" : 386
    "Cats" : 85
    "Rats" : 15
\`\`\`

`,
        }}
      />,
    )
    asFragment = component.asFragment
  })

  // @ts-ignore
  expect(asFragment()).toMatchSnapshot()
})

test('blockquote', async () => {
  let asFragment

  await act(() => {
    const component = render(
      <TestRenderer
        article={{
          ...article,
          body: `

And code blocks:

<blockquote>
i am a block
</blockquote>

<blockquote>
multi

line

block
</blockquote>

`,
        }}
      />,
    )
    asFragment = component.asFragment
  })

  // @ts-ignore
  expect(asFragment()).toMatchSnapshot()
})

test('reference fallback', async () => {
  let asFragment

  await act(() => {
    const component = render(
      <TestRenderer
        article={{
          ...article,
          body: `
[hello][hello]

![foobar][foobar]
`,
        }}
      />,
    )
    asFragment = component.asFragment
  })

  // @ts-ignore
  expect(asFragment()).toMatchSnapshot()
})

test('footnote', async () => {
  let asFragment

  await act(() => {
    const component = render(
      <TestRenderer
        article={{
          ...article,
          body: `
Hello[^hello]

[^hello]: i am a footnote
`,
        }}
      />,
    )
    asFragment = component.asFragment
  })

  // @ts-ignore
  expect(asFragment()).toMatchSnapshot()
})

test('multiline footnote', async () => {
  let asFragment

  await act(() => {
    const component = render(
      <TestRenderer
        article={{
          ...article,
          body: `
Hello[^hello]

[^hello]: i am a footnote
here is more
and more

this is outside of the footnote
`,
        }}
      />,
    )
    asFragment = component.asFragment
  })

  // @ts-ignore
  expect(asFragment()).toMatchSnapshot()
})

test('multiline indented footnote', async () => {
  let asFragment

  await act(() => {
    const component = render(
      <TestRenderer
        article={{
          ...article,
          body: `
Hello[^hello]

[^hello]: i am a footnote
    here is more
    **and more**

    \`and more\`

    and more

this is outside of the footnote
`,
        }}
      />,
    )
    asFragment = component.asFragment
  })

  // @ts-ignore
  expect(asFragment()).toMatchSnapshot()
})
