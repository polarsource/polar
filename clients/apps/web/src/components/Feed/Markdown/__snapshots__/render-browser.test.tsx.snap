// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XSS 1`] = `
<DocumentFragment>
  <div>
     

    <p>
      hello

    </p>
    <p>
      what

    </p>
    <br />
    <table>
      <thead>
        <tr>
          <th>
            TH
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            TD
          </td>
        </tr>
        <tr>
          <td>
            <p>
              <span>
                clickmep
              </span>
            </p>
          </td>
          <td>
            TD
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`basic 1`] = `
<DocumentFragment>
  <div>
    <h1
      id="h1"
    >
      h1
    </h1>
    <h2
      id="h2"
    >
      h2
    </h2>
    <h3
      id="h3"
    >
      h3
    </h3>
    <p>
      Hello 
      <strong>
        world
      </strong>
      !
    </p>
    <p>
      <a
        href="https://polar.sh/"
      >
        Polar
      </a>
    </p>
    <div
      class="my-4 flex flex-col items-center rounded-3xl bg-gray-100 px-8 py-4 text-center dark:bg-polar-700"
    >
      <p>
        This section is for premium subscribers only. Subscribe to 
        <a
          class="text-blue-500 hover:text-blue-400 dark:text-blue-400 dark:hover:text-blue-300"
          href="/pydantic/subscriptions"
        >
          Pydantic
        </a>
         to get access to it.
      </p>
    </div>
    <p>
      This is a normal 
      <strong>
        block
      </strong>
       of text 
      <a
        href="https://polar.sh/"
      >
        Polar
      </a>
       with 
      <em>
        various
      </em>
       formatting.
And here it continues in the same block.
    </p>
    <p>
      This is a different block.
      <br />
      With a linebreak! (double whitespace)
    </p>
    <blockquote>
      <p>
        This is a quoute!
      </p>
    </blockquote>
  </div>
</DocumentFragment>
`;

exports[`blockquote 1`] = `
<DocumentFragment>
  <div>
    <p>
      And code blocks:
    </p>
    <blockquote>
      i am a block

    </blockquote>
    <blockquote>
      <p>
        multi
      </p>
      <p>
        line
      </p>
      <p>
        block
      </p>
    </blockquote>
  </div>
</DocumentFragment>
`;

exports[`code 1`] = `
<DocumentFragment>
  <div>
     

    <p>
      code:::
    </p>
    <div
      class="relative my-2 w-full"
    >
      <div
        class="not-prose"
      >
        <pre
          style="text-align: left; white-space: pre; word-spacing: normal; word-break: normal; word-wrap: normal; color: rgb(0, 25, 58); background: rgb(243, 244, 247); tab-size: 4; font-size: .8em; border-radius: 2em; hyphens: none; overflow: auto; position: relative; margin: 0.5em 0px; padding: 2em;"
        >
          <code>
            <div
              class="m-0"
            >
              <span
                style="margin-right: 1.5rem; opacity: 0.2; font-size: .7rem; user-select: none; text-align: right; display: inline-block; width: 3.010299956639812px;"
              >
                1
              </span>
              <span
                style="color: rgb(0, 0, 255); font-style: italic;"
              >
                func
              </span>
              <span>
                 
              </span>
              <span
                style="color: rgb(124, 77, 255);"
              >
                main
              </span>
              <span
                style="color: rgb(57, 173, 181);"
              >
                (
              </span>
              <span
                style="color: rgb(57, 173, 181);"
              >
                )
              </span>
              <span>
                 
              </span>
              <span
                style="color: rgb(57, 173, 181);"
              >
                {
              </span>
            </div>
            <div
              class="m-0"
            >
              <span
                style="margin-right: 1.5rem; opacity: 0.2; font-size: .7rem; user-select: none; text-align: right; display: inline-block; width: 3.010299956639812px;"
              >
                2
              </span>
              <span
                style="color: rgb(57, 173, 181);"
              >
                }
              </span>
            </div>
          </code>
        </pre>
      </div>
      <button
        class="font-normal ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 whitespace-nowrap text-blue-500 dark:text-polar-200 hover:bg-blue-50 dark:hover:bg-polar-600 hover:border-blue-100 border dark:border-polar-600 border-blue-100 flex items-center justify-center p-2 absolute right-6 top-6 h-8 w-8 rounded-full bg-gray-50 text-sm dark:bg-gray-900"
      >
        <div
          class="flex flex-row items-center"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
            data-testid="ContentPasteOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 2h-4.18C14.4.84 13.3 0 12 0S9.6.84 9.18 2H5c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2m-7 0c.55 0 1 .45 1 1s-.45 1-1 1-1-.45-1-1 .45-1 1-1m7 18H5V4h2v3h10V4h2z"
            />
          </svg>
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`mermaid 1`] = `
<DocumentFragment>
  <div>
    <p>
      we love charts
    </p>
    <div
      class="animate-pulse rounded-md bg-muted w-full p-8 font-mono text-sm"
    >
      Initializing Mermaid Renderer...
    </div>
  </div>
</DocumentFragment>
`;

exports[`posts 1`] = `
<DocumentFragment>
  <div>
    <p>
      <img
        alt="Hero image"
        class="cursor-pointer"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2FGitHub%2520Repo-N4Yx3M434zkSTK3qQHubUyVNwXDohp.png&w=1920&q=90"
        srcset="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2FGitHub%2520Repo-N4Yx3M434zkSTK3qQHubUyVNwXDohp.png&w=750&q=90 1x, /_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2FGitHub%2520Repo-N4Yx3M434zkSTK3qQHubUyVNwXDohp.png&w=1920&q=90 2x"
        style="color: transparent;"
        width="672"
      />
    </p>
    <p>
      You can now seamlessly offer subscribers on Polar access to one, two, three... or countless private GitHub repositories ðŸŽ‰ This opens up and streamlines unlimited possibilities, monetization- and funding models.
    </p>
    <p>
      Open source developers can use it to offer:
    </p>
    <ul>
      <li>
        Access to private GitHub discussions & issues for sponsors
      </li>
      <li>
        Early access to new feature development before upstream push
      </li>
      <li>
        Sponsorware
      </li>
      <li>
        Premium educational materials & code
      </li>
    </ul>
    <p>
      Indie developers can use it to sell:
    </p>
    <ul>
      <li>
        Self-hosting products
      </li>
      <li>
        Courses
      </li>
      <li>
        Starter kits
      </li>
      <li>
        Open core software
      </li>
    </ul>
    <p>
      We're also the merchant of record and take care of VAT.
    </p>
    <p>
      Paired with premium 
      <a
        href="https://polar.sh/polarsource/posts/polar-creator-platform-for-open-source-developers"
      >
        Posts & Newsletters
      </a>
      , 
      <a
        href="https://polar.sh/polarsource/posts/upsell-discord-invites"
      >
        Discord invites
      </a>
       and 
      <a
        href="https://polar.sh/polarsource/posts/automate-sponsor-logos-ads"
      >
        Ads
      </a>
       open source developers can easily offer an unparalleled sponsorship & membership offering. To building an independent commercial open source initiative. You can even build a developer-first SaaS offering.
    </p>
    <p>
      Of course, using our API & SDK all of these capabilities can easily be integrated & upsold on your own site or service too.
    </p>
    <p>
      Polar is built by developers, for developers, and 
      <a
        href="https://github.com/polarsource/polar"
      >
        open source
      </a>
      . We want to empower a new generation of independent developers pursuing their passions full-time. So subscribe to get future product updates & posts as we'll write a series diving deep into all of the opportunities Polar can help facilitate for developers.
    </p>
    <p>
      Or better yet: Start building with Polar today - 
      <a
        href="https://polar.sh"
      >
        signup now
      </a>
      .
    </p>
    <iframe
      allowfullscreen=""
      class="my-2 aspect-[560/315] w-full max-w-full"
      frameborder="0"
      src="https://www.youtube.com/embed/QKx4o0z-SVY"
    />
    <h2
      id="lets-create-a-polar-subscription-benefit-for-a-github-repository"
    >
      Let's create a Polar subscription benefit for a GitHub repository
    </h2>
    <p>
      Let's rig it up and see how it works.
    </p>
    <ol>
      <li>
        Goto 
        <code>
          Subscriptions
        </code>
         in your dashboard and then click 
        <code>
          Edit tier
        </code>
         on your default 
        <strong>
          Free
        </strong>
         tier (for newsletters) to try it out
      </li>
      <li>
        Click 
        <code>
          New Benefit
        </code>
         (See below)
      </li>
    </ol>
    <p>
      <img
        alt="Create new benefit"
        class="cursor-pointer"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-ILn7PLAZL2eck6O9rKHCJIO4Nlgd4U.png&w=1920&q=90"
        srcset="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-ILn7PLAZL2eck6O9rKHCJIO4Nlgd4U.png&w=750&q=90 1x, /_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-ILn7PLAZL2eck6O9rKHCJIO4Nlgd4U.png&w=1920&q=90 2x"
        style="color: transparent;"
        width="672"
      />
    </p>
    <ul>
      <li>
        Select 
        <code>
          GitHub Repository Access
        </code>
         as the benefit type
      </li>
      <li>
        Select the 
        <code>
          Organization
        </code>
         the private repository belongs to or connect an organization (See notes below)
      </li>
    </ul>
    <p>
      <strong>
        Why organization vs. personal account?
      </strong>
       GitHub limits personal accounts to only invite collaborators without controlling their permissions, e.g committing to 
      <code>
        main
      </code>
      , pushing new releases etc. Organizations can set more restrictive permissions, e.g read-only. Therefore, due to the security risks, we've put support for personal accounts with this benefit behind a feature flag - reach out if you want it enabled anyway.
    </p>
    <p>
      <strong>
        Important pricing note:
      </strong>
       Collaborators are considered a seat on paid plans and you'll be billed accordingly. We therefore request read-only access to your organizations billing plan to warn about this in our UI so that you can price your Polar subscription to cover it.
    </p>
    <ul>
      <li>
        Click 
        <code>
          Install
        </code>
         to install the required Polar GitHub App
      </li>
    </ul>
    <p>
      <img
        alt="Create benefit - Install GitHub App"
        class="cursor-pointer"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-h5l5CsSZ3SI7LBg4Mx5fzokNzjpE6S.png&w=1920&q=90"
        srcset="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-h5l5CsSZ3SI7LBg4Mx5fzokNzjpE6S.png&w=750&q=90 1x, /_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-h5l5CsSZ3SI7LBg4Mx5fzokNzjpE6S.png&w=1920&q=90 2x"
        style="color: transparent;"
        width="672"
      />
    </p>
    <p>
      You'll be redirected to the GitHub App installation flow to grant access to select repositories. You can choose one of them, a select few or all of them.
    </p>
    <p>
      <strong>
        Important note:
      </strong>
       Selected repositories here are not automatically granted to subscribers on Polar. You'll explicitly select the one (1) desired repository to grant access to for the given subscription benefit you create on Polar. You're in complete control. Of course, you can create multiple benefits to grant access to numerous repositories, but they're all added explicitly.
    </p>
    <p>
      <img
        alt="GitHub installation flow"
        class="cursor-pointer"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-0GzjfOkdRqj0QyA07DqFIRstDbsouo.png&w=1920&q=90"
        srcset="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-0GzjfOkdRqj0QyA07DqFIRstDbsouo.png&w=750&q=90 1x, /_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-0GzjfOkdRqj0QyA07DqFIRstDbsouo.png&w=1920&q=90 2x"
        style="color: transparent;"
        width="672"
      />
    </p>
    <p>
      <em>
        Bonus: The Polar GitHub App will also enable you to use the issue crowdfunding & contributor rewards features with Polar too ðŸ”¥. They require read/write to issues & pull requests while administration is required for managing invites to desired GitHub repositories. Unfortunately, GitHub does not offer progressive permissions for Apps.
      </em>
    </p>
    <p>
      Once the Polar GitHub App has been installed you'll be taken back to Polar to continue the setup of the subscription benefit.
    </p>
    <ol>
      <li>
        Select the private 
        <code>
          Repository
        </code>
         you want to grant subscribers access to
      </li>
      <li>
        Select the desired GitHub 
        <code>
          Role
        </code>
        , e.g 
        <code>
          Read
        </code>
         for read access only.
      </li>
      <li>
        Click 
        <code>
          Create
        </code>
      </li>
    </ol>
    <p>
      <img
        alt="Select repository & role"
        class="cursor-pointer"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-MenUSRBNEXfw5ZmFkt5uP6Ffsh0I3q.png&w=1920&q=90"
        srcset="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-MenUSRBNEXfw5ZmFkt5uP6Ffsh0I3q.png&w=750&q=90 1x, /_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-MenUSRBNEXfw5ZmFkt5uP6Ffsh0I3q.png&w=1920&q=90 2x"
        style="color: transparent;"
        width="672"
      />
    </p>
    <p>
      We can now review our updates to the 
      <em>
        Free
      </em>
       tier and see our added GitHub Repository Access benefit we call 
      <code>
        Cutting edge
      </code>
      . We can also update our tier description to celebrate the new benefit.
    </p>
    <p>
      <img
        alt="Review tier"
        class="cursor-pointer"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-Z1UfL9ilQFHM64fMqIM3G6vI51bNSv.png&w=1920&q=90"
        srcset="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-Z1UfL9ilQFHM64fMqIM3G6vI51bNSv.png&w=750&q=90 1x, /_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-Z1UfL9ilQFHM64fMqIM3G6vI51bNSv.png&w=1920&q=90 2x"
        style="color: transparent;"
        width="672"
      />
    </p>
    <p>
      Looks great. Let's click 
      <code>
        Save Tier
      </code>
       and try it out as a subscriber.
    </p>
    <h2
      id="subscribers-accessing-github-repository"
    >
      Subscribers accessing GitHub repository
    </h2>
    <p>
      <img
        alt="Subscribe"
        class="cursor-pointer"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-sGIzRl0if9LmJd9lWhaGDpd6bLnsn3.png&w=1920&q=90"
        srcset="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-sGIzRl0if9LmJd9lWhaGDpd6bLnsn3.png&w=750&q=90 1x, /_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-sGIzRl0if9LmJd9lWhaGDpd6bLnsn3.png&w=1920&q=90 2x"
        style="color: transparent;"
        width="672"
      />
    </p>
    <p>
      Your audience can now see the benefit and easily subscribe to gain access to it. In case of email subscriptions, we'll prompt them to connect a GitHub account and immediately create the repository invitation once they have.
    </p>
    <p>
      Subscribers can see & manage their subscription benefits under 
      <code>
        My Subscriptions
      </code>
      .
    </p>
    <p>
      Look at that, they now see the 
      <code>
        Cutting edge
      </code>
       repository with a convenient activation link to it.
    </p>
    <p>
      <em>
        Note: They also receive an email invitation from GitHub.
      </em>
    </p>
    <p>
      <img
        alt="Gain access to repo via subscriptions"
        class="cursor-pointer"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-RxC5k4B9QPNxTGwYZUXhsjpjzKxgEw.png&w=1920&q=90"
        srcset="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-RxC5k4B9QPNxTGwYZUXhsjpjzKxgEw.png&w=750&q=90 1x, /_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-RxC5k4B9QPNxTGwYZUXhsjpjzKxgEw.png&w=1920&q=90 2x"
        style="color: transparent;"
        width="672"
      />
    </p>
    <p>
      So let's click the 
      <code>
        Goto polardemo/cutting-edge
      </code>
       link. Since it's our first time, we're prompted to accept the invitation first.
    </p>
    <p>
      <img
        alt="Accept invitation"
        class="cursor-pointer"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-qyWQF7WGTkld5JzcbZmvAYjibjspE7.png&w=1920&q=90"
        srcset="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-qyWQF7WGTkld5JzcbZmvAYjibjspE7.png&w=750&q=90 1x, /_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-qyWQF7WGTkld5JzcbZmvAYjibjspE7.png&w=1920&q=90 2x"
        style="color: transparent;"
        width="672"
      />
    </p>
    <p>
      ... and that's it. We're in!
    </p>
    <p>
      <img
        alt="Subscriber seeing the repository on GitHub"
        class="cursor-pointer"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-RMwsZSbqgWak7MfKjLjLPD8f25U3jq.png&w=1920&q=90"
        srcset="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-RMwsZSbqgWak7MfKjLjLPD8f25U3jq.png&w=750&q=90 1x, /_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-RMwsZSbqgWak7MfKjLjLPD8f25U3jq.png&w=1920&q=90 2x"
        style="color: transparent;"
        width="672"
      />
    </p>
    

    <h2
      id="manage-collaborators"
    >
      Manage collaborators
    </h2>
    <p>
      <img
        alt="image.png"
        class="cursor-pointer"
        data-nimg="1"
        decoding="async"
        height="100"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-Ihx3RL0EIobi7KlULBBvfeDYleStKA.png&w=1920&q=90"
        srcset="/_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-Ihx3RL0EIobi7KlULBBvfeDYleStKA.png&w=750&q=90 1x, /_next/image?url=https%3A%2F%2F7vk6rcnylug0u6hg.public.blob.vercel-storage.com%2Fimage-Ihx3RL0EIobi7KlULBBvfeDYleStKA.png&w=1920&q=90 2x"
        style="color: transparent;"
        width="672"
      />
    </p>
    <p>
      You can of course review and manage your collaborators on GitHub. As seen above, our subscriber is now a collaborator with 
      <code>
        Read
      </code>
       permissions. Just like we wanted.
    </p>
    <p>
      Oh, and of course, Polar will automatically remove collaborators once their subscription is no longer valid, i.e cancelled.
    </p>
    <hr />
    <p>
      In this example, we've setup seamless invitations to one (1) GitHub repository. However, it's the same process to add additional benefits for more repositories. So there's no limit to how creative you can get with your offerings - we can't wait to see some of your examples!
    </p>
    <p>
      Finally, make sure to subscribe below to stay up-to-date as we ship more product updates & write a series going deeper into different models Polar can help facilitate.
    </p>
    <p>
      Until next time, keep building!
    </p>
  </div>
</DocumentFragment>
`;
